### ISSUE TITLE
dead_code lint incorrectly warns about struct fields that aren't dead due to transmuting

### ISSUE THREAD

**Issue by kballard** [rust-lang/rust#15102]
The `dead_code` lint was recently (#14696) modified to detect dead struct fields. Unfortunately, it warns about fields even when it can't be 100% certain they're dead.

One case where this happens was filed as #15101, where the struct is used as C FFI. That case has the (undocumented) solution of using `#[repr(C)]` (even though `#[repr(C)]` hasn't been implemented to actually do anything yet).

Another case that doesn't have a clear solution beyond turning off the warning is a struct that contains fields for the purpose of being memory-layout-compatible with another struct.

Notably, in kballard/rust-lua@be46e268c0, the file `lib.rs` defines 3 structs with identical layout. Two of the structs properly use all defined fields, but one of the structs does not use one field (beyond initialization). However, there are methods that use `men::transmute()` to convert between the struct types. If this method is used, then the allegedly-dead field becomes live, as the destination struct uses that field.

I recognize that this is a bit of an edge case, but this isn't the only one. Besides C FFI, a quick test demonstrates that a supposedly-unused struct field that is in fact printed via a `format!("{:?}", foo)` directive is also considered dead. I wouldn't be surprised to find out that there are more edge cases either. The basic point here is that the `dead_code` lint is making an unqualified statement that the field is never used, but it doesn't actually have 100% certainty of that fact. It should be modified to recognize cases where the struct is used in ways the lint can't see into. So far the list of cases includes:
1. Being passed as a `*mut` to a C FFI function
2. Being passed to `mem::transmute()`
3. Being passed to `::debug::fmt::secret_poly`

And we can probably add casting `*mut StructType` to any `*mut T` (via `as`, e.g. `foo as *mut ()`) to the list.

/cc @jakub-

**Comment by steveklabnik** [comment]
Triage: here's a test case for the transmute issue:

```
#[derive(Debug)]
struct Foo {
    x: i32,
}

struct Bar {
    y: i32,
}

fn main() {
    let bar = Bar{ y: 4 };

    let foo: Foo = unsafe { std::mem::transmute(bar) };
    format!("{:?}", foo);
}
```

This does in fact warn about `y` but not `x`.

**Comment by ruud-v-a** [comment]
An other case where `dead_code` improperly warns about unused struct fields is if the field is of a type that implements `Drop`. It might be there to keep something alive even if it is not used otherwise, for instance to keep a file open. This should be fixable: suppress the warning if the field implements `Drop`.

**Comment by huonw** [comment]
@ruud-v-a the convention for locals that only care about the destructor works for struct fields too, that is, prefix it with an underscore:

```rust
struct Foo {
    _guard: GuardType
}
// similar to
let _guard = mutex.lock();
```

(https://github.com/rust-lang/rust/issues/21775 covers that situation specifically.)

**Comment by bltavares** [comment]
Triaging:

The following test case still warns on `y`, but not on `x`.

```rust
#[derive(Debug)]
struct Foo {
    x: i32,
}

struct Bar {
    y: i32,
}

fn main() {
    let bar = Bar{ y: 4 };

    let foo: Foo = unsafe { std::mem::transmute(bar) };
    format!("{:?}", foo);
}
```

Adding `#[repr(C)]` to the `Bar` struct does indeed remove the warning.

**Comment by Mark-Simulacrum** [comment]
As far as I can tell, the only remaining problem is that dead_code warns on fields that are used through transmute. However, `#[repr(C)]` fixes that; and using transmute without `#[repr(C)]` is a Bad Idea. Closing. Please reopen if I missed something.
